clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,237 @@
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.client.FilesetCatalog;
+import com.datastrato.gravitino.client.GravitinoClient;
+import com.datastrato.gravitino.client.GravitinoMetaLake;
+import com.datastrato.gravitino.file.Fileset;
+import com.datastrato.gravitino.shaded.com.google.common.base.Preconditions;
+import com.datastrato.gravitino.shaded.org.apache.commons.lang3.StringUtils;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.InvalidPathException;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.util.Progressable;
+
+public class GravitinoVirtualFileSystem extends FileSystem {
+  private FileSystem proxyFileSystem;
+  private Path workingDirectory;
+  private URI uri;
+  private NameIdentifier filesetIdentifier;
+
+  @Override
+  public void initialize(URI name, Configuration conf) throws IOException {
+    if (name.toString().startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX)) {
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(
+              conf.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_SERVER_URI_KEY)),
+          "Gravitino server uri is not set in the configuration");
+      this.filesetIdentifier = normalizedIdentifier(name.getPath());
+      String serverUri =
+          conf.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_SERVER_URI_KEY);
+      // TODO Need support more authentication types, now we only support simple auth
+      try (GravitinoClient client = GravitinoClient.builder(serverUri).withSimpleAuth().build()) {
+        GravitinoMetaLake metaLake =
+            client.loadMetalake(NameIdentifier.of(filesetIdentifier.namespace().level(0)));
+        Catalog catalog =
+            metaLake.loadCatalog(
+                NameIdentifier.of(
+                    filesetIdentifier.namespace().level(0),
+                    filesetIdentifier.namespace().level(1)));
+        FilesetCatalog filesetCatalog = (FilesetCatalog) catalog.asFilesetCatalog();
+        Fileset fileset = filesetCatalog.loadFileset(filesetIdentifier);
+        String storageLocation = fileset.storageLocation();
+        Preconditions.checkArgument(
+            storageLocation != null
+                && storageLocation.startsWith(
+                    GravitinoVirtualFileSystemConfiguration.HDFS_SCHEME_PREFIX),
+            "Storage location is not valid in the fileset metadata");
+        URI storageUri = new Path(storageLocation).toUri();
+        // close cache, so that the independent fileset uses the independent fs
+        conf.set(
```

Comment:
Should we add some logs to let user know the restriction? Could we add the restriction in the document?

Reply:
Yes, I think it is possible to add usage and limitations in the documentation